### DISCUSSION TITLE
"next even at ..." as output

### DISCUSSION THREAD

**Discussion by vtacquet** [jensrossbach/node-red-contrib-chronos#74] [Ideas & Requests]
following up on my original request (which was perfectly executed by the way !), is it in any way possible to retrieve the "next event" text as an optional output so I could put this on my dashboard for all users (and not only in the editor only for me) ? now it is kind of unknown for users when the next event will occur.
![CleanShot 2022-01-10 at 09 03 21](https://user-images.githubusercontent.com/5128967/148734345-5c4d0bd5-db48-445d-8a19-c2077331a9b7.png)
![CleanShot 2022-01-10 at 09 04 34](https://user-images.githubusercontent.com/5128967/148734417-99a2fcdb-c661-4985-ab9d-811a15656d52.png)

**Comment by jensrossbach** [comment]
Hi,
this should be possible as the event times are anyway calculated in the runtime and then just passed over to the editor. I could make the time available as a property for the JSONata based output types.

But please be aware that the time is available earliest with triggering of the first event as before there won't be any output produced.

**Comment by jensrossbach** [comment]
I had some more thoughts about this request and I came to the conclusion that an output message is not really appropriate for providing the time value. Firstly because of what I already mentioned above, i.e. that you won't receive the time before the first event triggers (so for instance, directly after a flow deploy, the time would not be available). And secondly, the calculation of the next trigger time takes place _after_ the output message has been sent. This means at the time when the output message is produced, the next trigger time is still the time of the current event (which doesn't help much).

Alternatively, the scheduler node could write the next trigger time to a context variable every time it is calculated. But then the name of the variable and the type of the context store either has to be configurable through the editor UI or the context store would need be hard-coded and the name would need to be a cryptic one based on the node's ID (in order to have it unique). While a hard-coded name is definitely not very nice, the other approach requires adding another setting to the UI. How important is such feature for you?

**Comment by vtacquet** [comment]
The current problems is that I have no idea when the flow will start because sun time events change every day, so currently I would have to go to the editor to see what the next event would be. But I understand the issue with redeploying and only seeing the next event when at least 1 event has triggered. Would it be a better way to add an additional output that sends info when a "query" input payload (next to toggle, reload, trigger) is sent ?

![CleanShot 2022-01-23 at 12 04 10](https://user-images.githubusercontent.com/5128967/150675457-a8e31a5d-3ac2-4b59-acd2-d8d43ebeadc8.png)

For now I have created another flow with offset events 5 minutes before the real action and delay it 4 minutes (because I can't select on 1 minute intervals) to notify us an event will be triggered, otherwise it scares the crap out of us when all shutters open/close without warning. 

![CleanShot 2022-01-23 at 12 06 13](https://user-images.githubusercontent.com/5128967/150675504-94077b6d-92b7-4f8c-94fa-c54fbbb406d8.png)

BTW is there a reason why the offset fields are rounded to 5 minutes ?

![CleanShot 2022-01-23 at 12 08 43](https://user-images.githubusercontent.com/5128967/150675579-40057a2e-d0d3-4e87-be0b-3362050fae13.png)

**Comment by jensrossbach** [comment]
> Would it be a better way to add an additional output that sends info when a "query" input payload (next to toggle, reload, trigger) is sent ?

This is something I would definitely like to avoid as it makes the node more complex and harder to understand (people would need to know that some ports are for the events and another one is for the time). In this case I would prefer the way through a context variable because it keeps the output ports "clean".

> BTW is there a reason why the offset fields are rounded to 5 minutes ?

The fields are not rounded to 5 minutes, you can enter any arbitrary integer number within the allowed range. Only the up/down buttons use a step of 5 minutes in order to be able to faster scroll up or down.

**Reply by jensrossbach** [reply]
Hi @vtacquet,

finally I decided to follow your proposal and implemented the possibility to activate a dedicated output port which emits messages with information about the next scheduled time (in `msg.payload` is the time of the upcoming event and in `msg.events` are the times of all configured events). The timestamps are in standard Node-RED timestamp format (numbers with milliseconds since Unix epoch), you can use the time change node to convert these numbers into a human readable format.

**Reply by vtacquet** [reply]
Hey Jens, thanks for the update !

Any idea why this test-node is sending 5 times the same message ? 

![CleanShot 2022-04-12 at 11 18 13](https://user-images.githubusercontent.com/5128967/162926466-7b9798b8-c666-4ca0-8075-534645f77f9c.png)
![CleanShot 2022-04-12 at 11 18 23](https://user-images.githubusercontent.com/5128967/162926479-bb494d92-c035-4092-a813-1dd9e4ecee1b.png)
![CleanShot 2022-04-12 at 11 18 32](https://user-images.githubusercontent.com/5128967/162926488-186515a4-f621-4eb0-9f15-e1385ff2c870.png)

**Reply by jensrossbach** [reply]
Hi, normally I would expect 4 messages as you have 4 schedule events configured. The message on the new output port is sent whenever any of the configured schedule events updates its time. On deploying a flow for instance, all configured schedule events are calculated and hence there will be x messages sent out (where x is the number of events). I will need to check why there is an additionaly one, I anyway already identified a possible optimization. I will also think if there is further optimization possible for avoiding duplicate messages of same content.

**Reply by vtacquet** [reply]
Hey Jens, why do you think it should send 4 messages, I understood from the doc that it would only send 1 message where msg.payload contains the time of the next event and msg.events contains next events 
![CleanShot 2022-04-13 at 08 19 39](https://user-images.githubusercontent.com/5128967/163112934-12fb599a-6a77-4b1b-8a6f-d865793de6c3.png)

**Reply by jensrossbach** [reply]
It sends one message (containing the upcoming schedule time as well as times of all events) everytime *any* of the configured nodes is updated. There are situation (e.g. deploy of a flow), where *all* events are updated at the same time (because the scheduled time is initially calculated for all events). And then, the same message is sent multiple times (as many as there are events configured).

**Reply by jensrossbach** [reply]
Hi @vtacquet, I just published version **1.17.1** which contains some improvements for emitting the event time updates. Now it should not send duplicate messages anymore.

Regards, Jens

**Reply by vtacquet** [reply]
Hey Jens, I don't get any output to the dedicated output port anymore, not when changing the schedule and not with the "reload" input payload 
![CleanShot 2022-04-19 at 15 17 29](https://user-images.githubusercontent.com/5128967/164012871-c461ae3b-a793-460e-9d70-27e7f358a068.png)

**Reply by jensrossbach** [reply]
Hi, can you please describe the exact steps to reproduce? Maybe I missed some use cases during testing, sorry.

**Reply by jensrossbach** [reply]
The problem with "reload" especially is that it does not change the schedule times, so with my new implementation, this prevents the message from being sent. So I must invalidate the last message upon reload so that sending of it forced even though it does not contain new content.

**Reply by jensrossbach** [reply]
Hi @vtacquet, please try out release 1.17.2.

**Reply by vtacquet** [reply]
very nice @jensrossbach, output works fine again !

**Comment by vtacquet** [comment]
Hi Jens,

When you set it to your own value and you click Done, after reopening the node, it is changed to the nearest value from the list. See video below.

https://user-images.githubusercontent.com/5128967/150679618-e725da59-4152-4b3b-b97c-8ba188fe0c91.mp4

Vince

**Comment by jensrossbach** [comment]
ok, in this case it is a bug, can you please enter an issue in the bug tracker?
